worker/credentialvalidator/worker.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package credentialvalidator
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+var logger = loggo.GetLogger("juju.api.credentialvalidator")
+
+// ErrChanged indicates that a Worker has bounced because its
+// Check result has changed.
+var ErrChanged = errors.New("cloud credential validity has changed")
+
+// ErrModelCredentialChanged indicates that a Worker has bounced
+// because model credential was replaced.
+var ErrModelCredentialChanged = errors.New("unexpected model credential")
+
+// ErrModelDoesNotNeedCredential indicates that a Worker has been uninstalled
+// since the model does not have a cloud credential set as the model is
+// on the cloud that does not require authentication.
+var ErrModelDoesNotNeedCredential = errors.New("model is on the cloud that does not need auth")
+
+// Predicate defines a predicate.
+type Predicate func(base.StoredCredential) bool
+
+// Facade exposes functionality required by a Worker to access and watch
+// a cloud credential that a model uses.
+type Facade interface {
+
+	// ModelCredential gets model's cloud credential.
+	// Models that are on the clouds that do not require auth will return
+	// false to signify that credential was not set.
+	ModelCredential() (base.StoredCredential, bool, error)
+
+	// WatchCredential gets cloud credential watcher.
+	WatchCredential(string) (watcher.NotifyWatcher, error)
+}
+
+// IsValid returns true when the given credential is valid.
+func IsValid(c base.StoredCredential) bool {
+	return c.Valid
+}
+
+// Config holds the dependencies and configuration for a Worker.
+type Config struct {
+	Facade Facade
+	Check  Predicate
+}
+
+// Validate returns an error if the config cannot be expected to
+// drive a functional Worker.
+func (config Config) Validate() error {
+	if config.Facade == nil {
+		return errors.NotValidf("nil Facade")
+	}
+	if config.Check == nil {
+		return errors.NotValidf("nil Check")
+	}
+	return nil
+}
+
+// New returns a Worker that tracks the result of the configured
+// Check on the Model's cloud credential, as exposed by the Facade.
+func New(config Config) (*Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	modelCredential, exists, err := config.Facade.ModelCredential()
```

Comment:
These steps should all be at the top of the loop method (before the for{})
Besides following current practice, it means that modelCredential and watcher don't need to be worker attributes. The watcher would be added to the catacomb when created in the loop.

Reply:
Getting model credential is the biggest and not a simple operation. There is no need to get it all the time only if a valid and interesting change is reported.
The same approach is used in migration flag which to me was a better and very similar example to what I was doing than any other worker we have.

Reply:
it won't be got ll the time - it's just moving the code from here to the top of the worker loop (not inside the for). 99% of workers have all their logic inside the loop(). Migration worker would be the outlier here. It's better to keep all the worker logic together in the loop function.

Reply:
While it has been idiomatic in other places to have the New method just initialise the structure and start the loop, I don't see anything wrong with the initialisation including getting the credentials from the DB.

Reply:
Consistency is good. If there's no real reason to stray, why introduce the cognitive overhead. Code is write once, read many.

Reply:
Plus the suggested approach removes the need for unnecessary worker attributes. Less LOC = less bugs by definition.

Reply:
I think consistency is good. Ideally new people coming into the code will be able to see the patterns of how things are written. I'd agree that moving the code into the loop function for consistency is worth it.